renter/session.go

```diff
@@ -0,0 +1,219 @@
+// account's balance can be used as the payment method for other RPC calls.
+func (s *Session) FundAccount(accountID types.PublicKey, amount types.Currency, payment PaymentMethod) (types.Currency, error) {
+	if payment == nil {
+		return types.ZeroCurrency, errors.New("payment method is required")
+	} else if _, ok := payment.(*payByContract); !ok {
+		return types.ZeroCurrency, errors.New("ephemeral accounts must be funded by a contract")
+	}
+
+	stream, err := s.session.DialStream()
+	if err != nil {
+		return types.ZeroCurrency, fmt.Errorf("failed to open new stream: %w", err)
+	}
+	defer stream.Close()
+
+	settings, err := s.currentSettings()
+	if err != nil {
+		return types.ZeroCurrency, errors.New("price table invalid or expired")
+	}
+
+	err = rpc.WriteRequest(stream, rhp.RPCFundAccountID, &settings.UID)
+	if err != nil {
+		return types.ZeroCurrency, fmt.Errorf("failed to write account balance request: %w", err)
+	}
+
+	if err := payment.Pay(stream, settings.RPCFundAccountCost.Add(amount)); err != nil {
+		return types.ZeroCurrency, fmt.Errorf("failed to pay for account balance: %w", err)
+	}
+
+	err = rpc.WriteResponse(stream, &rhp.RPCFundAccountRequest{
+		AccountID: accountID,
+	}, nil)
+	if err != nil {
+		return types.ZeroCurrency, fmt.Errorf("failed to write account balance request: %w", err)
+	}
+
+	var resp rhp.RPCFundAccountResponse
+	if err := rpc.ReadResponse(stream, &resp); err != nil {
+		return types.ZeroCurrency, fmt.Errorf("failed to read account balance response: %w", err)
+	}
+
+	return resp.Balance, nil
+}
+
+// Settings returns the current settings for the host. If a payment method is
+// provided the setting UID should be used when interacting with the host.
+func (s *Session) Settings(payment PaymentMethod) (settings types.HostSettings, _ error) {
+	stream, err := s.session.DialStream()
+	if err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to open stream: %w", err)
+	}
+	defer stream.Close()
+
+	if err := rpc.WriteRequest(stream, rhp.RPCSettingsID, nil); err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to write settings request: %w", err)
+	}
+
+	var resp rhp.RPCSettingsResponse
+	if err := rpc.ReadResponse(stream, &resp); err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to read settings response: %w", err)
+	}
+
+	if err := json.Unmarshal(resp.Settings, &settings); err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to decode settings: %w", err)
+	}
+
+	if payment == nil {
+		settings.UID = [16]byte{}
+		return
+	} else if err := payment.Pay(stream, settings.RPCHostSettingsCost); err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to pay for settings: %w", err)
+	}
+
+	if err = rpc.ReadResponse(stream, &rhp.RPCTrackingResponse{}); err != nil {
+		return types.HostSettings{}, fmt.Errorf("failed to read tracking response: %w", err)
+	}
+
+	s.settingsMu.Lock()
+	s.settings = settings
+	s.settingsExpiration = time.Now().Add(settings.Validity)
```

Comment:
computing `time.Now().Add(settings.Validity)` should be done as soon as possible after receiving the settings